added support for sht4x

ndml28 opened the pull request:
Included SHT4x Sensor. Was tested on a SAMD21 MCU.

ndml28 committed:
added support for sht4x

ndml28 committed:
added support for sht4x

winkj commented on SHTSensor.h:
```diff
@@ -201,14 +204,14 @@ class SHTI2cSensor : public SHTSensorDriver {
   SHTI2cSensor(uint8_t i2cAddress, uint16_t i2cCommand, uint8_t duration,
                float a, float b, float c,
-               float x, float y)
+               float x, float y, float z, uint8_t CMD_Size)
```
since CMD_Size is just a parameter, it should use camel case (cmdSize)

winkj commented on SHTSensor.h:
```diff
@@ -226,6 +229,8 @@ class SHTI2cSensor : public SHTSensorDriver {
   float mX;
   float mY;
+  float mZ;
+  uint8_t mCMD_Size;
```
same here, please use camel case (mCmdSize)

winkj commented on SHTSensor.cpp:
```diff
@@ -203,6 +203,55 @@ class SHT3xSensor : public SHTI2cSensor
+  SHT4xSensor(uint8_t i2cAddress = SHT4X_I2C_ADDRESS_44)
+      : SHTI2cSensor(i2cAddress, SHT4X_ACCURACY_HIGH,
+                     SHT4X_ACCURACY_HIGH_DURATION,
+                     -45, 175, 65535, 125, 65535,-6,1)
```
why is cmdSize 1 here? 0xFD00, 0xF600, 0xE000 et al are all two bytes?

ndml28 replied:
The reason for the cmd size varible was that the SHT4x commands are in fact only one byte. The function readSample() from the SHTI2CSensor base class however used the hard coded CMD Size of two byte. So in order to make the function more flexible I introduced the variable. I defined the commands as two byte commands (i.e. 0xFD00) because they are also shifted in the readSample function. This way the older sensors and the SHT4x can use the same functions and that was in my opinion the easiest implementation where not too much has to be changed. If you have any other ideas please let me know

winkj replied:
ah, makes sense. I think we'll probably want to readSample() down the line, but I think that can be a refactoring later

winkj reviewed (commented):
Thanks for the merge request. what was the reason to introduce a variable cmd size? from what I can see, the SHT4x cammands are two bytes as well

winkj commented on SHTSensor.cpp:
```diff
@@ -133,12 +133,12 @@ bool SHTI2cSensor::readSample()
 
   val = (data[3] << 8) + data[4];
-  mHumidity = mX * (val / mY);
+  mHumidity = mZ + mX * (val / mY);
```
I feel the names for the parameters would read more easily if they followed the same pattern as humidity, i.e. mX + mY * (val / mZ). Thoughts?

winkj commented on SHTSensor.h:
```diff
@@ -201,14 +204,14 @@ class SHTI2cSensor : public SHTSensorDriver {
    * and the values `x' and `y' to convert the fixed-point humidity value
    * received by the sensor to a floating point value using the formula:
-   * humidity = x * (rawHumidity / y)
+   * humidity = z + x * (rawHumidity / y)
```
see above; I think it would be more logical to do x + y * (rawHumidity / z)

winkj commented on SHTSensor.cpp:
```diff
@@ -148,7 +148,7 @@ class SHTC1Sensor : public SHTI2cSensor
     SHTC1Sensor()
         // clock stretching disabled, high precision, T first
-        : SHTI2cSensor(0x70, 0x7866, 15, -45, 175, 65535, 100, 65535)
+        : SHTI2cSensor(0x70, 0x7866, 15, -45, 175, 65535, 100, 65535,0,2)
```
same comment as above regarding order: would make more sense to read 
`SHTI2cSensor(0x70, 0x7866, 15, -45, 175, 65535, 0, 100, 65535)` 

(i.e. move 0 up in the argument list; cmd lengths likely not needed)

Also, please match the coding style regarding spaces after commas

winkj commented on SHTSensor.cpp:
```diff
@@ -176,7 +176,7 @@ class SHT3xSensor : public SHTI2cSensor
       : SHTI2cSensor(i2cAddress, SHT3X_ACCURACY_HIGH,
                      SHT3X_ACCURACY_HIGH_DURATION,
-                     -45, 175, 65535, 100, 65535)
+                     -45, 175, 65535, 100, 65535,0,2)
```
See above

MBjoern reviewed (commented):
Also mention the SHT4x support in the README file in the Supported sensors section.

ndml28 commented:
I will include the propsed changes.

ndml28 committed:
changed to camelCase, adjusted order of variables for humidity calculation

ndml28 committed:
changed to camelCase, adjusted order of variables for humidity calculation

winkj commented:
@rnestler @MBjoern Code looks good in my opinion. I don't have any hardware available for testing right now, can you build and test?

winkj committed:
fix compiler warning in SHTSensor::setAccuracy()

SHTSensor::setAccuracy(SHTAccuracy newAccuracy) is simply returning
false in the base class, since most of the sensors don't support
it. as such, 'newAccuracy' is newer used.

Mark it as unused to get rid of the compiler warning in restrictive
compilation setups

winkj committed:
check working sensor communication in init()

Previously, init() behaved differently depending on
whether SHTSensor was instantiated with an fixed sensor
type or not (auto-detect). In the auto-detect case, it would
trigger a 'readSample()' and thus ensure that the
sensor was communicating, while the fixed initialization
no such check was performed.

This change addressed this by calling readSample() in both cases,
and thus guaranteeing that if init() returns true, the sensor driver
was successfully created, and communication with the sensor on the bus
has been established

rnestler commented:
> @rnestler @MBjoern Code looks good in my opinion. I don't have any hardware available for testing right now, can you build and test?

I have a SHT4x on flex here, but would need to organize / build an adapter to wire it up to the Arduino.

MBjoern commented:
Hi there

Thanks for the added support of the SHT4x sensor. There are 2 issues remaining from my tests:

1. Auto detection didn't work with the SHT4x. Did you test this on your end? Check the autodection example of the library.

2. The sensor readings without auto detection didn't work either. There was a successful initialization of the sensor, but after that, no successful readings. This is related to issue 10: https://github.com/Sensirion/arduino-sht/issues/10. After applying the suggested fix the readings succeed. @winkj Could you followup on the issue 10 and adjust it if applicable?

As soon as issue 10 has been fixed, you can tackle the autodetection issue.

winkj committed:
Use non-GCC specific way to avoid unused warning

winkj committed:
release I2C bus after sending read command

Wire::endTransmission() has an optional argument that
defines whether the bus is released or not. At the time
of writing the library, we set it to hold the bus. This
causes issues on other platforms.

winkj commented:
FYI, a PR for Issue #10 has been submitted and should hit the master branch soon

winkj committed:
remove redundant readSample() call

Since init() now calls readSample() also
for SHTSensor instances created with a fixed type,
we can remove the additional check in the
auto-detect loop

MBjoern commented:
Ok Issue 10 has been merged.

@ndml28 Please rebase your pull request and have a look at the auto detection feature, if you don't mind. Then we'll be able to merge it to master.

MBjoern reviewed (dismissed):
Rebase branch and fix autodetection for SHT4x to work properly.

ndml28 commented:
Sounds good! 
I'll do that.
I also have some hardware to test it with on my side.

MBjoern commented:
@ndml28 Sounds great, We're excited to test it in action soon.

ndml28 committed:
Added Delay for AUTODETECT

ndml28 committed:
Merge branch 'sht4x' of github.com:ndml28/arduino-sht into sht4x

ndml28 commented:
So the auto detection feature indeed behaves a bit weird. What fixed it for me with the SHT4X was to add a delay. 40 ms and it works fine (I tried smaller values but no success). I don't know exactly why it is required since the minimum wait time between I2C commands on the SHT4X is only 1 ms according to the datasheet. So it might not be the cleanest solution. I could not try it with the other sensors since I only have a board with the SHT4X on hand.

MBjoern reviewed (commented):
I've tested with SHT4x and SHT3x and it works perfectly! Looks good to me
@winkj  Could you do a final code check and give the final approval to merge?

winkj commented:
I can't see any negative impact of the delay.

@ndml28 the last commit (delay for auto detect) added a bunch of extra whitespace... would you mind redoing this commit? Feel free to just force push into the PR branch

ndml28 committed:
rm whitespace

winkj reviewed (approved):
Looks good to me, thanks for the update.

I will go ahead and merge this!

winkj commented:
I manually rebased these changes on top of master to get rid of the merge commit, and went squashed it into a single commit. All code changes have now been merged.

@ndml28 Thanks a lot for implementing this, and apologies for letting this sit for so long!

winkj commented:
Merged in 1de2e9a039c7c8f2e4d87f13cfaa5197d09937ca
